sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/jsonExpressions.scala

```diff
@@ -747,8 +748,13 @@ case class StructsToJson(
 
 object JsonExprUtils {
 
-  def validateSchemaLiteral(exp: Expression): StructType = exp match {
-    case Literal(s, StringType) => CatalystSqlParser.parseTableSchema(s.toString)
+  def validateSchemaLiteral(exp: Expression): DataType = exp match {
+    case Literal(s, StringType) =>
+      try {
+        DataType.fromJson(s.toString)
```

Comment:
I don't think we should support JSON format here. DDL formatted schema is preferred. JSON in functions.scala is supported for backword compatibility because SQL functions wasn't added first.

Reply:
I believe we should support JSON format because:
- Functionality of SQL and Scala (and other languages) DSL should be equal otherwise we push users to use Scala DSL because SQL has less features.
- The feature allows to save/restore schema in JSON format. Customer's use case is to have data in JSON format + meta info including schema in JSON format too. Schema in JSON format gives them more opportunities for processing in programatic way. 
- For now JSON format give us more flexibility and allows `MapType` (and `ArrayType`) as the root type for result of `from_json`

Reply:
Usually they should be consistent but we don't necessarily support the obsolete functionality newly and consistently. I'm not sure how common it is to write the JSON literal as a schema via SQL. How do they get the metadata and how do they insert it into SQL? Is that the only way to do it?

Reply:
> How do they get the metadata ...

Metadata is stored together with data in distributed fs and loaded by a standard facilities of language. 

> and how do they insert it into SQL?

SQL statements are formed programmatically as  strings, and loaded schemas are inserted in particular positions in the string ( you can think about it as quasiquotes in Scala). The formed sql statements are sent via JDBC to Spark.

> Is that the only way to do it?

Probably it is possible to convert schemas in JSON format to DDL format but:
- it requires much more effort and time than just modifying 5 lines proposed in the PR
- Schema in DDL supports only `StructType` as root types. It is not possible to specify `MapType` like in the test: https://github.com/apache/spark/blob/master/sql/core/src/test/scala/org/apache/spark/sql/JsonFunctionsSuite.scala#L330-L345

Reply:
> Shall we add the support with type itself with CatalystSqlParser.parseDataType too?

I will do but it won't solve customer's problem fully.

> Also, are you able to use catalogString?

I just check that:
```scala
val schema = MapType(StringType, IntegerType).catalogString
val ds = spark.sql(
      s"""
        |select from_json('{"a":1}', '$schema')
      """.stripMargin)
ds.show()
```
and got this one:

```
extraneous input '<' expecting {'SELECT', 'FROM', ...}(line 1, pos 3)

== SQL ==
map<string,int>
---^^^
; line 2 pos 7
```

The same with ` val schema = new StructType().add("a", IntegerType).catalogString
`

```
== SQL ==
struct<a:int>
------^^^
; line 2 pos 7
org.apache.spark.sql.AnalysisException
```

Am I doing something wrong?

Reply:
I mean, type string support by `CatalystSqlParser.parseDataType` (like array<...> or struct<...>) into `from_json`. This will support `struct<a:int>` if I am not mistaken.

Reply:
I mean, like what we do in Python side:

https://github.com/apache/spark/blob/4f1e38649ebc7710850b7c40e6fb355775e7bb7f/python/pyspark/sql/types.py#L808-L814

Reply:
If possible, I like @HyukjinKwon 's approach. I remember correctly we just keep json schema formats for back-compatibility. In future major releases, I think we possibly drop the support.

Reply:
@maropu @HyukjinKwon Please, have a look at the PR: https://github.com/apache/spark/pull/21550